### PR TITLE
Disallow partial selection of entities

### PR DESCRIPTION
**Pull request by atimmer**
## Summary

This PR can be summarized in the following changelog entry:

* [SnippetEditor] Disallow partial selection of replacement variables.

## Relevant technical choices:

We don't want to allow partial selection of entities because that results
in very unexpected behavior. Because we cannot distinguish between mouse
and keyboard input we have to compromise.

When moving into an entity with the keyboard now the whole entity will be
selected. When selecting half-way into an entity the whole entity will be
selected.

## Test instructions

This PR can be tested by following these steps:

* Go into the editor and try selecting entities half-way.
* They should now be selected completely.

Fixes https://github.com/Yoast/wordpress-seo/issues/10064.

### CODE REVIEW

File: composites/Plugin/SnippetEditor/selection.js [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+		 * If the previous selection contained the entity and we now have it half-way
+		 * selected that means we need to move the selection completely outside the
+		 * entity.
+		 */
+		if ( selectionContainsEntity( previousSelection, entityRange ) ) {
+			selection = selection.merge( {
+				[ endOffsetProperty ]: startOfEntity,
+			} );
+
+			/*
```
Comment (IreneStr) [inline composites/Plugin/SnippetEditor/selection.js]: Indentation of comments

File: composites/Plugin/SnippetEditor/selection.js [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+ * @param {EditorState} editorState The current state of the editor.
+ * @param {EditorState} previousEditorState The state before the onChange event.
+ * @returns {EditorState} The new state of the editor.
+ */
+export function selectReplacementVariables( editorState, previousEditorState ) {
+	const selection = editorState.getSelection();
+	const previousSelection = previousEditorState.getSelection();
+	const contentState = editorState.getCurrentContent();
+
+	// We are fixing the selection, but if it didn't change there is nothing to fix.
```
Comment (IreneStr) [inline composites/Plugin/SnippetEditor/selection.js]: 'Fixing' is not very clear + add comma after 'change'

File: composites/Plugin/SnippetEditor/selection.js [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+	if ( selection === previousSelection ) {
+		return editorState;
+	}
+
+	const newSelection = alterSelection( selection, previousSelection, contentState );
+
+	// If we always force a new selection we will trigger onChange events indefinitely.
```
Comment (IreneStr) [inline composites/Plugin/SnippetEditor/selection.js]: Please add comma after 'new selection'

File: composites/Plugin/SnippetEditor/tests/selectionTest.js [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+ * - `before` is the selection state before the editor `onChange`.
+ * - `after` is the selection state after/during the editor `onChange`.
+ * - `expected` is the selection state the function should result in.
+ */
+
+/**
+ * Converts a visualizable text to a DraftJS SelectionState.
+ *
+ * @param {string} text The text to convert.
+ * @param {string} key The key of the block the selection is on.
```
Comment (IreneStr) [inline composites/Plugin/SnippetEditor/tests/selectionTest.js]: blockKey

File: composites/Plugin/SnippetEditor/tests/selectionTest.js [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+
+	it( "selects the whole entity when selecting a small part from the left", () => {
+		expectToMatch( {
+			before: "Text []entity entity Text",
+			after: "Text [e]ntity entity Text",
+			expected: "Text [entity] entity Text",
+		} );
+	} );
+
+	it( "moves over an entity to the right", () => {
```
Comment (IreneStr) [inline composites/Plugin/SnippetEditor/tests/selectionTest.js]: The it doesn't describe the expected selection

File: composites/Plugin/SnippetEditor/tests/selectionTest.js [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+		expectToMatch( {
+			before: "Text []entity entity Text",
+			after: "Text e[]ntity entity Text",
+			expected: "Text [entity] entity Text",
+		} );
+	} );
+
+	it( "moves over an entity to the left", () => {
```
Comment (IreneStr) [inline composites/Plugin/SnippetEditor/tests/selectionTest.js]: se above

File: composites/Plugin/SnippetEditor/tests/selectionTest.js [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+
+	it( "selects multiple entities when selecting half-way", () => {
+		expectToMatch( {
+			before: "[]Text entity entity Text",
+			after: "Text ent[ity ent]ity Text",
+			expected: "Text [entity entity] Text",
+		} );
+	} );
+
+	it( "moves out of selection when pressing arrow keys", () => {
```
Comment (IreneStr) [inline composites/Plugin/SnippetEditor/tests/selectionTest.js]: Or clicking before/after it. We don't know whether it's done by the arrow keys or the mouse/

File: composites/Plugin/SnippetEditor/selection.js [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+/**
+ * Determines the entity key at a certain position.
+ *
+ * @param {ContentState} contentState The current content.
+ * @param {string}       blockKey     The block key.
+ * @param {number}       position     The position to get the entity for.
+ * @returns {?number} Entity key or null.
+ */
+export function getEntityAtPosition( contentState, blockKey, position ) {
+	const block = contentState.getBlockForKey( blockKey );
```
[the remark this answers is not in the document]
Comment (atimmer) [reply composites/Plugin/SnippetEditor/selection.js]: I do actually. When calling `block.getEntityAt`.

File: composites/Plugin/SnippetEditor/selection.js [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+ * It is important to understand what the difference is between the anchor and
+ * the focus in a selection.
+ *
+ * @link https://draftjs.org/docs/api-reference-selection-state.html#start-end-vs-anchor-focus
+ * @link https://developer.mozilla.org/en-US/docs/Web/API/Selection
+ *
+ * @param {boolean} isBackward Whether the selection is done backwards.
+ * @returns {Object} The start and end offset properties.
+ */
+let getOffsetProperties = function( isBackward ) {
```
Comment (igorschoester) [inline composites/Plugin/SnippetEditor/selection.js]: This should be `const`.

File: composites/Plugin/SnippetEditor/selection.js [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+ * @param {SelectionState} previousSelection The selection before the onChange event.
+ * @param {ContentState}   contentState      The current content state.
+ * @returns {SelectionState} The new selection state.
+ */
+function alterSelection( selection, previousSelection, contentState ) {
+	// DraftJS handles settings these properties to the correct value.
+	let startOffset = selection.getStartOffset();
+	let startBlock = selection.getStartKey();
+	let endOffset = selection.getEndOffset();
+	let endBlock = selection.getEndKey();
```
Comment (igorschoester) [inline composites/Plugin/SnippetEditor/selection.js]: These 4 could be `const` I think.

File: composites/Plugin/SnippetEditor/selection.js [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+
+	let { startOffsetProperty, endOffsetProperty } = getOffsetProperties( selection.getIsBackward() );
```
Comment (igorschoester) [inline composites/Plugin/SnippetEditor/selection.js]: And these could be `const`.

File: composites/Plugin/SnippetEditor/selection.js [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+
+	return selection.getStartOffset() <= startOfEntity &&
+	       selection.getEndOffset()   >= endOfEntity;
+}
+
+/**
+ * Alters the selection to include the entity if the selection is inside the
+ * entity.
+ *
+ * @param {SelectionState} selection         The current selection.
```
Comment (igorschoester) [inline composites/Plugin/SnippetEditor/selection.js]: selectionState?

File: composites/Plugin/SnippetEditor/selection.js [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+
+/**
+ * Selects a replacement variable completely when the selection is inside it.
+ *
+ * @param {EditorState} editorState The current state of the editor.
+ * @param {EditorState} previousEditorState The state before the onChange event.
+ * @returns {EditorState} The new state of the editor.
+ */
+export function selectReplacementVariables( editorState, previousEditorState ) {
+	const selection = editorState.getSelection();
```
Comment (igorschoester) [inline composites/Plugin/SnippetEditor/selection.js]: selectionState?

File: composites/Plugin/SnippetEditor/tests/selectionTest.js [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+ * We're using the brackets here to make it easier to visualize where the selection
+ * is and should be at certain points.
+ *
+ * - `before` is the selection state before the editor `onChange`.
+ * - `after` is the selection state after/during the editor `onChange`.
+ * - `expected` is the selection state the function should result in.
+ */
+
+/**
+ * Converts a visualizable text to a DraftJS SelectionState.
```
Comment (igorschoester) [inline composites/Plugin/SnippetEditor/tests/selectionTest.js]: It also applies the selection to the given state.

**Comment by Xyfi** [comment]
CR :+1:

**Comment by Xyfi** [comment]
Acceptance :construction: 

An error is thrown when the editor is opened, and the entities are not properly shown, when holding backspace the last character can't be removed and the mentions are not shown when typing `%`. 

The error when opening the editor:
screen shot 2018-06-22 at 11 39 05

A gif demonstrating a few of the problem occuring:
pr 607-error

**Comment by atimmer** [comment]
Fixed that.

**Comment by Xyfi** [comment]
CR :+1:  Acceptance :+1:

**Pull request merged**